[GTK][WPE] Support asynchronously returning values from user script messages

siwei-li opened the pull request:
#### 9267215ef0d1b1b2f0233ff63940fa293eb51d30
<pre>
[GTK][WPE] Support asynchronously returning values from user script messages
<a href="https://bugs.webkit.org/show_bug.cgi?id=243492">https://bugs.webkit.org/show_bug.cgi?id=243492</a>

Reviewed by Carlos Garcia Campos.

Expose WebKit&apos;s functionality of responding asynchronously
to user script messages in the public API, by means of a
returned Promise object.

A new helper class &quot;WebKitUserScriptMessgeReply&quot; is added to
Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp;
it holds the returned JS value or the error message from executing
the user script message, which will be sent to the user by the
&quot;didPostMessageWithAsyncReply()&quot; method in the &quot;ScriptMessageClientGtk&quot;
class.

If asynchronous handling is wished for, the new
&quot;webkit_user_content_manager_register_script_message_handler_with_reply()&quot;
method should be used to register the message handler, its associated signal
being &quot;WebKitUserContentManager::script-message-with-reply-received&quot;.
The user should ideally provide a callback reply in the script message,
otherwise an undefined value will be returned.

* Source/WebKit/Shared/API/APISerializedScriptValue.h:
* Source/WebKit/UIProcess/API/glib/APISerializedScriptValueGLib.cpp:
(API::SerializedScriptValue::createFromJSCValue):
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
(_WebKitScriptMessageReply::_WebKitScriptMessageReply):
(_WebKitScriptMessageReply::sendValue):
(_WebKitScriptMessageReply::sendErrorMessage):
(_WebKitScriptMessageReply::~_WebKitScriptMessageReply):
(webkit_script_message_reply_ref):
(webkit_script_message_reply_unref):
(webKitScriptMessageReplyCreate):
(webkit_script_message_reply_return_value):
(webkit_script_message_reply_return_error_message):
(webkit_user_content_manager_register_script_message_handler):
(webkit_user_content_manager_register_script_message_handler_with_reply):
(webkit_user_content_manager_register_script_message_handler_in_world):
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManagerPrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
(UserScriptMessageTest::UserScriptMessageTest):
(UserScriptMessageTest::~UserScriptMessageTest):
(UserScriptMessageTest::registerHandler):
(UserScriptMessageTest::scriptMessageReceived):
(UserScriptMessageTest::waitUntilMessageReceived):
(UserScriptMessageTest::asyncScriptMessageReceived):
(UserScriptMessageTest::waitUntilPromiseResolved):
(UserScriptMessageTest::runAsyncJavaScriptFinished):
(UserScriptMessageTest::postMessageAndWaitForPromiseResolved):
(testUserContentManagerScriptMessageWithReplyReceived):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/257016@main">https://commits.webkit.org/257016@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/619c995f05eee7ea061ff3e0a08ee32a7f860862

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97512 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6767 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30677 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107016 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167280 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101469 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7103 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35523 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89906 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103685 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103167 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5319 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84143 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32324 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87192 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88983 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75249 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/769 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/757 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21923 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4831 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5570 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44395 "Found 3 new test failures: fast/forms/file/entries-api/image-no-transcode-open-panel.html, fast/images/animated-heics-draw.html, fast/images/animated-heics-verify.html (failure)") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2007 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41297 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -69,8 +69,80 @@ struct _WebKitUserContentManagerPrivate {
+    {
+        JSCContext* jscContext = jsc_context_new();
+        JSValueRef jsValueRef = serializedScriptValue.deserialize(jscContextGetJSContext(jscContext), nullptr);
+        message = jscValueCreate(jscContext, jsValueRef);
```
Here we should use the internal utility functions we have to create or reuse a shared context used for deserializing JS values. You can see an example in `WebKitJavascriptResult.cpp`. It would be like this:

```suggestion
        auto* jsContext = SharedJavascriptContext::singleton().getOrCreateContext();
        message = jscContextGetOrCreateValue(jsContext, serializedScriptValue.deserialize(jscContextGetJSContext(jscContext), nullptr));
```

The important bit is using the `getOrCreateContext()` method: it's somewhat expensive to create new JS execution contexts, so here using the shared really helps. The rest is the same you had&mdash;it looks a bit nicer to me as a single line, but there was nothing wrong with your approach.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -86,12 +158,12 @@ static void webkit_user_content_manager_class_init(WebKitUserContentManagerClass
+     * webkit_user_content_manager_register_script_message_handler_with_reply()
      *
      * Since: 2.8
      */
```
This is modifying the documentation for the existing signal. Please add a new one below, which can be a copy-paste of the existing one with the modifications you did here and with `Since: 2.38`.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -104,6 +176,16 @@ static void webkit_user_content_manager_class_init(WebKitUserContentManagerClass
+
+    signals[SCRIPT_REQUEST_RECEIVED] =
+        g_signal_new(
+            "script--received",
```
I suppose this is a typo, there are two dashes in the signal name and we had agred to call the new signal `script-request-received`... probably you removed the “request” word accidentally 😉 

```suggestion
            "script-request-received",
```

mcatanzaro replied:
My $0.02: the `script-message` terminology is slightly nicer than `script-request`.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +306,60 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+ *
+ * Since: 2.38
+ */
+JSCValue* webkit_user_script_message_get_message(WebKitUserScriptMessage* userScriptMessage)
```
Now that I see this function name again, I am not super convinced anymore about the name, because it has the “message” word twice in it. The alternative that comes to mind is `webkit_user_script_message_get_value()` but it feels a bit vague... maybe @mcatanzaro or @carlosgcampos have better naming ideas?

mcatanzaro replied:
`webkit_user_script_message_get_value()` sounds good to me. It's literally a `JSCValue`, after all.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +306,60 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+JSCValue* webkit_user_script_message_get_message(WebKitUserScriptMessage* userScriptMessage)
+{
+    if (userScriptMessage->message == nullptr)
+        return jsc_value_new_undefined(jsc_value_new());
```
Let's also use the `.getOrCreateContext()` helper here.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +306,60 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+ */
+void webkit_user_script_message_reply(WebKitUserScriptMessage* message, JSCValue* replyValue)
+{
+    message->replyValue = replyValue;
```
Like you did above, we always add `g_return[_val]_if_fail()` calls at the top of public API functions to do some checks on the passed values. For example here we would want to check the following:

```cpp
    g_return_if_fail(message != nullptr);
    g_return_if_fail(JSC_IS_VALUE(replyValue));
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +306,60 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+ */
+void webkit_user_script_message_error(WebKitUserScriptMessage* message, GError* error)
+{
+    message->replyError.reset(error);
```
Same here, let's use `g_return_if_fail()` to check that both `message` and `error` are non-null.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -242,18 +372,22 @@ class ScriptMessageClientGtk final : public WebScriptMessageHandler::Client {
     {
+        WebKitUserScriptMessage* message = webKitUserScriptMessageCreate(serializedScriptValue, completionHandler);
+        g_signal_emit(m_manager, signals[SCRIPT_REQUEST_RECEIVED], m_handlerName, message);
+        webkit_javascript_result_unref(message);
```
This is using the incorrect unref function for `WebKitUserScriptMessage`, should be:

```suggestion
        webkit_user_script_message_unref(message);
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -320,6 +454,36 @@ void webkit_user_content_manager_unregister_script_message_handler(WebKitUserCon
+ *
+ * Registers a new user script message handler in script world with name @world_name.
+ * If NULL is passed as the world_name, the default world will be used.
+ * See webkit_user_content_manager_register_script_message_handler() for full description.
```
Reminder for a followup patch: when we deprecate the old functions, we may want to move the longer description here, and write `See webkit_user_content_manager_register_script_message_handler_with_reply() instead` in the deprecated ones.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -320,6 +454,36 @@ void webkit_user_content_manager_unregister_script_message_handler(WebKitUserCon
+    if (worldName == nullptr)
+        Ref<WebScriptMessageHandler> handler = WebScriptMessageHandler::create(makeUnique<ScriptMessageClientGtk>(manager, name, true), AtomString::fromUTF8(name), API::ContentWorld::pageContentWorld());
+    else
+        Ref <WebScriptMessageHandler> handler = WebScriptMessageHandler::create(makeUnique<ScriptMessageClientGtk>(manager, name, true), AtomString::fromUTF8(name), webkitContentWorld(worldName));
```
This won't build. The code below uses the `handler` variable outside of the `if`-`else` statement, which means that it needs to be defined outside from it:

```cpp
    Ref<WebScriptMessageHandler> handler;
    if (worldName)
        handler = WebScriptMessageHandler::create(...);
    else
        handler = ...;

    return manager->priv->userContentController->addUserScriptMessageHandler(handler.get());

aperezdc commented on Source/WebKit/UIProcess/API/gtk/WebKitUserContentManager.h:
```diff
@@ -118,6 +118,19 @@ webkit_user_content_manager_remove_filter_by_id                        (WebKitUs
+
+WEBKIT_API void
+webkit_user_script_message_unref          (WebKitUserScriptMessage    *user_script_message);
+
```
The headers are missing the declarations for `webkit_user_script_message_get_message()`, `webkit_user_script_message_reply()`, and `webkit_user_script_message_error()`.

aperezdc reviewed (changes requested):
The first line of the commit log is missing the `[GTK][WPE]` tags, but other than a few minor comments (see my notes below) I like where this is going. We will need to add API tests for the new functions, those will go into `Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp`, which already has a `testUserContentManagerScriptMessageReceived()` test case which can be useful as guidance to write a new one.

One question for @carlosgcampos or @mcatanzaro: would we prefer to have the `WebKitUserScriptMessage` implementation moved inside `WebKitUserContent.cpp` instead? We have other boxed types there already 🤔

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -320,6 +454,36 @@ void webkit_user_content_manager_unregister_script_message_handler(WebKitUserCon
+    g_return_val_if_fail(WEBKIT_IS_USER_CONTENT_MANAGER(manager), FALSE);
+    g_return_val_if_fail(name, FALSE);
+
+    if (worldName == nullptr)
```
I'm not sure why the style EWS is not complaining about this, but WebKit code style is to avoid explicit checks against `nullptr`:

```suggestion
    if (!worldName)
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +433,74 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+        : completionHandler(completionHandler)
+        , referenceCount(1)
+    {
+        JSCContext* jscContext = jsc_context_new();
```
It looks like you forgot to change this to use `SharedJavaScriptContext::singleton().getOrCreateContext()`, like in the other part of the patch.

Also, there is something I missed in my previous review (sorry!): we should create an undefined value here if deserializing fails, see longer explanation below.

Actually, I think we do not even need to do that, because calling a message handler without passing a parameter with `window.webkit.messageHandler.<name>.postMessage()` is the same as passing `undefined` implicitly&mdash;we need to double check this.

aperezdc replied:
You marked this comment as resolved, but the changes in the PR are still using `jsc_context_new()` instead of `SharedJavaScriptContext::singleton().getOrCreateContext()`... maybe you forgot to push that change?

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +247,64 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+JSCValue* webkit_user_script_message_get_value(WebKitUserScriptMessage* userScriptMessage)
+{
+    if (!userScriptMessage->message)
+        return jsc_value_new_undefined(SharedJavaScriptContext::singleton().getOrCreateContext());
```
The function is annotated with `Return: (transfer none)`, which indicates that the caller does _not_ need to take care of doing an unref on the returned value. But this returns a new value, with a reference count of `1`, and if the caller never unrefs it, the value will be leaked.

We should make sure inside the `_WebKitUserScriptMessage()` constructor that `userScriptMessage->message` will always be a valid object instead.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +247,64 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+    if (!userScriptMessage->message)
+        return jsc_value_new_undefined(SharedJavaScriptContext::singleton().getOrCreateContext());
+
+    return userScriptMessage->message;
```
This is a smart pointer, to return the raw `WebKitUserScriptMessage*` pointer we need to call `.get()` on it:

 ```suggestion
    return userScriptMessage->message.get();
```

aperezdc reviewed (changes requested):
I did a couple of comments, of which one of them I had missed before. Please take a look.

Also, you may want to rebase your changes on top of the current status of the `main` branch. Recently we merged a patch which unifies the headers for both the GTK and WPE ports, and you will not need to duplicate the addition of the new functions in two different headers 😉

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.h.in:
```diff
@@ -116,6 +116,30 @@ webkit_user_content_manager_remove_filter_by_id                        (WebKitUs
+
+WEBKIT_API void
+webkit_user_script_message_error          (WebKitUserScriptMessage *user_script_message,
+                                           GErroe *error);
```
Here's a typo: `GErroe` → `GError`.

```suggestion
                                           GError *error);
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.h.in:
```diff
@@ -116,6 +116,30 @@ webkit_user_content_manager_remove_filter_by_id                        (WebKitUs
+webkit_user_script_message_reply          (WebKitUserScriptMessage *user_script_message,
+                                           JSCValue *reply_value);
+
+WEBKIT_API void
```
These new functions should be in `WebKitUserContent.h.in`, for consitency now that the implementation is in `WebKitUserContent.cpp`.

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -279,8 +279,7 @@ class UserScriptMessageTest : public WebViewTest {
     {
-        return worldName ? webkit_user_content_manager_register_script_message_handler_in_world(m_userContentManager.get(), handlerName, worldName)
-            : webkit_user_content_manager_register_script_message_handler(m_userContentManager.get(), handlerName);
+        return webkit_user_content_manager_register_script_message_handler_with_reply(m_userContentManager.get(), handlerName, worldName);
```
While technically this is okay, if we replace the existing implementation we no longer have a test for the old functions. We should add instead a `registerAsyncHandler()` function, or add a parameter to the existing one, for example:

```cpp
bool registerHandler(const char* handlerName, const char* worldName, bool async)
{
    if (async) {
        // Use the new function.
    } else {
        // The existing code here goes here, so we can continue testing the old functions.
    }
}

aperezdc reviewed (changes requested):
I think this is almost ready. Note that the code does not build due to a typo, maybe you forgot to amend the patch with some local changes?

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +437,122 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+ *
+ * A helper class that sends user provided script to the signal handler.
+ *
+ * Since: 2.38
```
We are a bit late in the 2.38 development cycle (we have branched a few weeks ago) so maybe we need to leave this new feature for 2.40&mdash;although I would love to have it in 2.38 and given it's a new addition it will definitely not break existing applications and the risk of cherry-picking this to the 2.38 branch should be very close to none.

@carlosgcampos WDYT, should we try to have this ready soon for 2.38, or do we leave for 2.40?

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -320,6 +348,34 @@ void webkit_user_content_manager_unregister_script_message_handler(WebKitUserCon
+ *
+ * Registering a script message handler will fail if the requested
+ * name has been already registered before.
+ *
```
Maybe I would also add a note here telling that handlers which were configured with this function can be unregistered with the existing `webkit_user_content_manager_unregister_script_message_handler()` and `webkit_user_content_manager_unregister_script_message_handler_in_world()` functions&mdash;maybe users will expect an `_unregister_script_message_handler_with_reply()` which does not exist, and mentioning the existing functions we can nudge them in the right direction.

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -438,6 +515,39 @@ static void testUserContentManagerScriptMessageInWorldReceived(UserScriptMessage
+
+    /*
+    test->postMessageAndWaitForPromiseResolved("msg", "");
+    */
```
This looks like some leftover test code, let's remove it before merging the patch 😉

aperezdc reviewed (changes requested):
Only a couple of things needed to get this past the finish line, it's very close now!

- There are a few [hints by the style bot](https://ews-build.webkit.org/#/builders/6/builds/88440), they seem easy to apply. Hint: you can use `Tools/Script/check-webkit-style -g HEAD` to run the style checks locally on the most recent Git commit (which is typically the one with your patch).
- The new test case seems to be crashing, as seen at the end of the “stdio” [log of the api-gtk builder](https://ews-build.webkit.org/#/builders/34/builds/67900).
- I wrote a couple of very minor comments.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -104,6 +105,28 @@ static void webkit_user_content_manager_class_init(WebKitUserContentManagerClass
+            "script-request-received",
+            G_TYPE_FROM_CLASS(gObjectClass),
+            static_cast<GSignalFlags>(G_SIGNAL_RUN_LAST | G_SIGNAL_DETAILED),
+            0, nullptr, nullptr,
```
The reason for having a `gboolean` return value from signal handlers is to allow returning `TRUE` from one to indicate that the script message was handled, and that no mode handlers should run. Reminder: there can be multiple handlers connected to the same signal, and only one can produce a result.

To ensure signal emission stops after a handler returns `TRUE`, we need to pass [g_signal_accumulator_true_handled](https://docs.gtk.org/gobject/func.signal_accumulator_true_handled.html) here:

```suggestion
            0, g_signal_accumulator_true_handled, nullptr,
```

A note in the signal documentation indicating that the first handler that returns `TRUE` results in no other handlers running would be welcome, too.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -242,18 +266,22 @@ class ScriptMessageClientGtk final : public WebScriptMessageHandler::Client {
+    void didPostMessageWithAsyncReply(WebPageProxy&, FrameInfoData&&, API::ContentWorld&, WebCore::SerializedScriptValue& serializedScriptValue, WTF::Function<void(API::SerializedScriptValue*, const String&)>&& completionHandler) override
     {
+        WebKitUserScriptMessage* message = webKitUserScriptMessageCreate(serializedScriptValue, WTFMove(completionHandler));
+        g_signal_emit(m_manager, signals[SCRIPT_REQUEST_RECEIVED], m_handlerName, message);
```
When signals have a return type, we need to pass one last parameter to `g_signal_emit()` to indicate where to put the returned value. In this case we don't need to do anything with it (we are only interested in the effect of the `g_signal_accumulator_true_handled` accumulator), but it still needs to be passed:

```suggestion
        gboolean resultValue;
        g_signal_emit(m_manager, signals[SCRIPT_REQUEST_RECEIVED], m_handlerName, message, &resultValue);
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +437,122 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+
+    GRefPtr<JSCValue> message;
+    GRefPtr<JSCValue> replyValue;
+    GUniquePtr<GError> replyError;
```
I think in the end we do not really need to save these. We can invoke the `completionHandler` directly inside `webkit_user_script_message_reply()` and `webkit_user_script_message_error()`.

aperezdc replied:
```suggestion
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +437,122 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+        JSCContext* jscContext = SharedJavascriptContext::singleton().getOrCreateContext();
+        JSValueRef jsValueRef = serializedScriptValue.deserialize(jscContextGetJSContext(jscContext), nullptr);
+        message = jscValueCreate(jscContext, jsValueRef);
+    }
```
We are missing a destructor here: if `completionHandler` has never been used, that means `webkit_user_script_message_reply()` or `webkit_user_script_message_error()` have _not_ been called. In that case, the destructor needs to construct an undefined JS value and send it as reply.

To know whether `completionHandler` has been used or not, we can reset it to `nullptr` after using it. Then, if `completionHandler` is non-null, that means it has not been used.

```suggestion
    }
    
    ~_WebKitUserScriptMessage()
    {
        if (completionHandler) {
            // Make a JSCValue for "undefined" and send it as the reply.
        }
    }
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +437,122 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+{
+    g_return_if_fail(message != nullptr);
+    g_return_if_fail(JSC_IS_VALUE(replyValue));
+    message->replyValue = replyValue;
```
If we go with the approach that I suggested above, we need to construct an `API::SerializedScriptValue` which represents the passed `replyValue`, call the `completionHandler`, and then set it to `nullptr` (to mark it as used/invalid). We can also add a check to make sure we get warnings if trying to use the `completionHandler` more than once:

```suggestion
    g_return_if_fail(JSC_IS_VALUE(replyValue));
    g_return_if_fail(message->completionHandler);

    auto serializedValue = API::SerializedScriptValue::create(...);
    message->completionHandler(serializedValue.get(), { });
    message->completionHandler = nullptr;
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +437,122 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+{
+    g_return_if_fail(message != nullptr);
+    g_return_if_fail(error != nullptr);
+    message->replyError.reset(error);
```
Same here: this needs to call the `completionHandler` with the error.

One possible way of converting the `GError` to a string can be:

```cpp
GUniquePtr<char> formattedError(g_strdup_printf("%s (%i): %s", g_quark_to_string(error->domain), error->code, error->message));

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -315,13 +319,68 @@ class UserScriptMessageTest : public WebViewTest {
+            }
+            webkit_javascript_result_unref(jsResult);
+        } else {
+            g_printerr("js result err: %s\n", error->message);
```
The test should always work correctly and this `else` block should never be reached. We can check that with:

```suggestion
            g_assert_not_reached();
```

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -438,6 +515,39 @@ static void testUserContentManagerScriptMessageInWorldReceived(UserScriptMessage
+    valueString.reset(WebViewTest::javascriptResultToCString(javascriptResult));
+    g_assert_cmpstr(valueString.get(), ==, "y");
+
+    valueString.reset(jsc_value_to_string(test->postMessageAndWaitForPromiseResolved("msg", "'42'")));
```
The value being sent back as result from the promise is the string `Inception` (line 332 above), so that's the value we should check for:

```suggestion
    valueString.reset(jsc_value_to_string(test->postMessageAndWaitForPromiseResolved("msg", "Inception")));
```

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -438,6 +515,39 @@ static void testUserContentManagerScriptMessageInWorldReceived(UserScriptMessage
+    g_assert_cmpstr(valueString.get(), ==, "42");
+
+    /*
+    test->postMessageAndWaitForPromiseResolved("msg", "");
```
It would be good to also test that using `webkit_user_script_message_error()` results in the `Promise` producing the expected error instead of a value. Another test we could add is using a script-world other than the default.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +438,142 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+/**
+ * WebKitUserScriptMessage: (ref-func webkit_user_script_message_ref) (unref-func webkit_user_script_message_unref)
+ *
+ * A helper class that sends user provided script to the signal handler.
```
This is not exactly a class. I would say something like: A script message reply. I'm not sure this belongs to WebKitUserContent, since it might be confused with WebKitUserScript, I think I would move this to WebKitUserContentManager since this is the reply to the script-message-received signal.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +438,142 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+ * Since: 2.40
+ */
+struct _WebKitUserScriptMessage {
+    _WebKitUserScriptMessage(WebCore::SerializedScriptValue& serializedScriptValue, WTF::Function<void(API::SerializedScriptValue*, const String&)>&& completionHandler)
```
The original API uses Function, but we can probably use CompletionHandler here wrapping the given function.

aperezdc replied:
TIL about what `WTF::CompletionHandler` exactly does, nice suggestion 👌🏼

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +438,142 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+        auto serializedValue = API::SerializedScriptValue::create(jscContextGetJSContext(jscContext), jscValueGetJSValue(value), nullptr);
+
+        completionHandler(serializedValue.get(), { });
+        completionHandler = nullptr;
```
With CompletionHandler we don't need this.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +438,142 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+        if (completionHandler) {
+            auto value = adoptGRef(jsc_value_new_undefined(jsc_value_get_context(message.get())));
+            sendValue(value.get());
+        }
```
This should b documented, if the user doesn't send a reply, an automatic reply with undefined value will be sent. Note that this is different to the mac behavior that replies with an error saying the user didn't send any reply. I'm not sure which approach is better.

aperezdc replied:
When I suggested sending an `undefined` I was thinking in trying to make the new API a replacement for the existing functionality, which is fire-and-forget. If you are okay with the idea, we could mark the old functions as deprecated.

Of course, this behaviour needs a comment in the documentation.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +438,142 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+    {
+        JSCContext* jscContext = SharedJavascriptContext::singleton().getOrCreateContext();
+        JSValueRef jsValueRef = serializedScriptValue.deserialize(jscContextGetJSContext(jscContext), nullptr);
+        message = jscValueCreate(jscContext, jsValueRef);
```
adoptGRef()

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +438,142 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+ * webkit_user_script_message_get_value:
+ * @user_script_message: A #WebKitUserScriptMessage
+ *
+ * Return the message sent in by user.
```
The message was not sent by the user, it was received from the js application.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +438,142 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+ * @user_script_message: A #WebKitUserScriptMessage
+ * @reply_value: Reply value of the provided script
+ *
+ * Provide the reply to the user provided script.
```
The user didn't provide the script. This is the reply to a script message received

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +438,142 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+
+    GUniquePtr<char> formattedError(g_strdup_printf("%s (%i): %s", g_quark_to_string(error->domain), error->code, error->message));
+    message->completionHandler(nullptr, String::fromUTF8(formattedError.get()));
+    message->completionHandler = nullptr;
```
We could add sendError and handle the completion handler details in the struct like sendValue does.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -433,3 +438,142 @@ API::ContentRuleList& webkitUserContentFilterGetContentRuleList(WebKitUserConten
+ *
+ * Since: 2.40
+ */
+void webkit_user_script_message_error(WebKitUserScriptMessage* message, GError* error)
```
I don't think GError is useful here, I think it would be simpler to receive an error message that is what the reply expects. Otherwise a script handling errors would need to understand the error quark and code to parse it.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.h.in:
```diff
@@ -150,6 +150,32 @@ webkit_user_content_filter_ref            (WebKitUserContentFilter     *user_con
 
+#define WEBKIT_TYPE_USER_SCRIPT_MESSAGE    (webkit_user_script_message_get_type())
+
+typedef struct _JSCValue JSCValue;
```
Can't we include <jsc/jsc.h>?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -104,6 +105,28 @@ static void webkit_user_content_manager_class_init(WebKitUserContentManagerClass
+     * @manager: the #WebKitUserContentManager
+     * @user_script_message: the #WebKitUserScriptMessage holding the value received from the JavaScript world.
+     *
+     * This signal is emitted when JavaScript in a web view calls
+     * <code>window.webkit.messageHandlers.<name>.postMessage()</code>, after registering
+     * <code><name></code> using
+     * webkit_user_content_manager_register_script_message_handler_with_reply()
```
We should document here that the callback is expected to send a reply, that could be handled asynchronously by taking a reference of the reply object and what happens if no reply is sent by the user.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -227,9 +250,10 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    ScriptMessageClientGtk(WebKitUserContentManager* manager, const char* handlerName)
+    ScriptMessageClientGtk(WebKitUserContentManager* manager, const char* handlerName, bool async)
```
async -> supportsAsyncReply

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -320,6 +349,38 @@ void webkit_user_content_manager_unregister_script_message_handler(WebKitUserCon
+ * Registers a new user script message handler in script world.
+ *
+ * Registers a new user script message handler in script world with name @world_name.
+ * If NULL is passed as the world_name, the default world will be used.
```
NULL -> %NULL
world_name -> @world_name

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -320,6 +349,38 @@ void webkit_user_content_manager_unregister_script_message_handler(WebKitUserCon
+ * webkit_user_content_manager_register_script_message_handler_with_reply:
+ * @manager: A #WebKitUserContentManager
+ * @name: Name of the script message channel
+ * @world_name: the name of a #WebKitScriptWorld
```
(nullable)

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -320,6 +349,38 @@ void webkit_user_content_manager_unregister_script_message_handler(WebKitUserCon
+ *
+ * Registers a new user script message handler in script world with name @world_name.
+ * If NULL is passed as the world_name, the default world will be used.
+ * See webkit_user_content_manager_register_script_message_handler() for full description.
+ *
+ * Registering a script message handler will fail if the requested
+ * name has been already registered before.
+ *
```
We should document here why this is different to register_script_message_handler, because it's pretty much the same documentation. We should explain that when using this to register the handler the signal emitted is script-message-with-reply received and that we are expected to reply to the script.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -438,6 +518,66 @@ static void testUserContentManagerScriptMessageInWorldReceived(UserScriptMessage
+    valueString.reset(jsc_value_to_string(test->postMessageAndWaitForPromiseResolved("msg", "'Inception'", "WebExtensionTestScriptWorld")));
+    g_assert_cmpstr(valueString.get(), ==, "Inception");
+    valueString.reset(jsc_value_to_string(test->postMessageAndWaitForPromiseResolved("msg", "throw 'err msg';", "WebExtensionTestScriptWorld", "err msg")));
+    g_assert_cmpstr(valueString.get(), ==, "err msg");
```
We should test the case of not replying to the script from the callback.

carlosgcampos reviewed (changes requested):
This looks great, but I have a few suggestions for the API. I'm sorry because I know this iterated several times and I already knew the proposal API, but after looking at the patch I find the new API a bit confusing. It's weird that script-request-received signal receives a WebKitUserScriptMessage object. It can also be confused with WebKitUserScript. So, I propose the following changes in the API:
- rename the signal as script-message-with-reply-received: and the callback doesn't receive a single object, it receives a WebKitJavascriptResult (like script-message-received) and WebKitScriptMessageReply. This makes more obvious that when usng register_script_message_handler_with_reply the signal that will be emitted is script-message-with-reply-received. And it's also more obvious that a reply is expected.
- Rename WebKitUserScriptMessage as WebKitScriptMessageReply: it doesn't handle the actual message, only the reply and can be used to reply with a js value or error message, so it could have webkit_script_message_reply_return_value and webkit_script_message_reply_retun_error_message, for example.

I'll add more inline comments.

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -315,13 +319,67 @@ class UserScriptMessageTest : public WebViewTest {
+            };
+            webkit_user_script_message_reply_return_value(userScriptMessageReply, jsValue.get());
+        }
+    return TRUE;
```
There is one level of indentation missing here.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.h.in:
```diff
@@ -116,6 +121,30 @@ webkit_user_content_manager_remove_filter_by_id                        (WebKitUs
+typedef struct _WebKitUserScriptMessageReply WebKitUserScriptMessageReply;
+
+WEBKIT_API GType
+webkit_user_script_message_reply_get_type       (void);
```
The opening parens for the function arguments should be all aligned vertically on the same column... I know: it's a bit of an odd choice, but that's what many projects (including WebKit) have inherited from GLib, which also formats public headers that way.

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -85,7 +87,6 @@ static bool isStyleSheetInjectedForURLAtPath(WebViewTest* test, const char* path
         : test->runJavaScriptAndWaitUntilFinished(kStyleSheetTestScript, &error.outPtr());
     g_assert_nonnull(javascriptResult);
     g_assert_no_error(error.get());
-
```
This line deletion does not need to be in the patch.

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -291,6 +294,7 @@ class UserScriptMessageTest : public WebViewTest {
 
     static void scriptMessageReceived(WebKitUserContentManager* userContentManager, WebKitJavascriptResult* jsResult, UserScriptMessageTest* test)
     {
+        g_signal_handlers_disconnect_by_func(userContentManager, reinterpret_cast<gpointer>(scriptMessageReceived), test);
```
This line is the same as the one immediately below, you can remove it.

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
```diff
@@ -438,6 +513,68 @@ static void testUserContentManagerScriptMessageInWorldReceived(UserScriptMessage
+    // Not replying to the script from callback should give us an undefined value.
+    valueString.reset(jsc_value_to_string(test->postMessageAndWaitForPromiseResolved("msg", "", "", "")));
+    g_assert_cmpstr(valueString.get(), ==, "undefined");
+    valueString.reset(jsc_value_to_string(test->postMessageAndWaitForPromiseResolved("msg", "", "WebExtensionTestScriptWorld", "err msg")));
```
There is this error in the EWS log:

```
ERROR:/app/webkit/Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:380:
static void UserScriptMessageTest::runAsyncJavaScriptFinished(GObject*, GAsyncResult*, UserScriptMessageTest*):
assertion failed (test->m_userScriptMessageErrorMessage.get() == error->message): (NULL == "TypeError: Not enough arguments")
```

This is caused because the second argument to the function (empty string) is placed in the JS snippet `window.webkit.messageHandlers.%s.postMessage(%s)` inside the parenthesis, which results in not passing any parameter:

```
window.webkit.messageHandlers.msg.postMessage()
```

For example we could pass `null` or `undefined` instead, to avoid having zero arguments:

```suggestion
    valueString.reset(jsc_value_to_string(test->postMessageAndWaitForPromiseResolved("msg", "undefined", "", "")));
    g_assert_cmpstr(valueString.get(), ==, "undefined");
    valueString.reset(jsc_value_to_string(test->postMessageAndWaitForPromiseResolved("msg", "undefined", "WebExtensionTestScriptWorld", "err msg")));
```

I believe the test case should start passing with this small change.

aperezdc reviewed (commented):
@siwei-li This is now very close to landing, it only need a tiny fix for a couple of test cases, and small touch ups (indentation). 

Thanks again for working on this and all the patience with all the reviews 🙏🏼

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.cpp:
```diff
@@ -20,9 +20,14 @@
+#include <jsc/JSCValuePrivate.h>
 #include <wtf/HashMap.h>
 #include <wtf/NeverDestroyed.h>
+#include <wtf/glib/GUniquePtr.h>
```
I guess the new includes are no longer needed here, right?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -104,6 +109,33 @@ static void webkit_user_content_manager_class_init(WebKitUserContentManagerClass
+    /**
+     * WebKitUserContentManager::script-message-with-reply-received:
+     * @manager: the #WebKitUserContentManager
+     * @user_script_message: the #WebKitUserScriptMessage holding the value received from the JavaScript world.
```
user_script_mssage -> script_message_reply

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -104,6 +109,33 @@ static void webkit_user_content_manager_class_init(WebKitUserContentManagerClass
+            0, g_signal_accumulator_true_handled, nullptr,
+            g_cclosure_marshal_generic,
+            G_TYPE_BOOLEAN, 1,
+            WEBKIT_TYPE_USER_SCRIPT_MESSAGE_REPLY);
```
I think the signal should receive both the message as WebKitJavaScriptResult and the reply

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +256,158 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
 }
 
+/**
+ * WebKitUserScriptMessageReply: (ref-func webkit_user_script_message_reply_ref) (unref-func webkit_user_script_message_reply_unref)
```
WebKitUserScriptMessageReply -> WebKitScriptMessageReply

siwei-li replied:
Just would like to confirm that I should refactor all `WebKitUserScriptMessageReply` into `WebKitScriptMessageReply`, along with the naming of related methods, correct? (The rest of the suggestions have mostly been resolved, and thank you so much for all the feedback!)

carlosgcampos replied:
Yes, if you have addressed all other review comments

carlosgcampos replied:
WebKitUserScriptMessageReply could be confusing with something related to WebKitUserScript and the signal name is script-message-with-reply-received (no user-script-message-with-reply-received) so it's consistent with signal name too.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +256,158 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+    {
+        JSCContext* jscContext = SharedJavascriptContext::singleton().getOrCreateContext();
+        JSValueRef jsValueRef = serializedScriptValue.deserialize(jscContextGetJSContext(jscContext), nullptr);
+        message = adoptGRef(jscValueCreate(jscContext, jsValueRef));
```
We don't need to handle this here, the message can be a WebKitJavaScriptResult like the signal with no reply does.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +256,158 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+        completionHandler(serializedValue.get(), { });
+    }
+
+    void sendErrMsg(const char* errMsg)
```
sendErrMsg -> sendErrorMessage
errMsg -> errorMessage

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +256,158 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+JSCValue* webkit_user_script_message_reply_get_value(WebKitUserScriptMessageReply* userScriptMessageReply)
+{
+    return userScriptMessageReply->message.get();
+}
```
We don't need this either

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +256,158 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+void webkit_user_script_message_reply_return_value(WebKitUserScriptMessageReply* message, JSCValue* replyValue)
+{
+    g_return_if_fail(message != nullptr);
+    g_return_if_fail(message->completionHandler);
```
We should probably document this can be called twice

aperezdc replied:
> We should probably document this can be called twice

**CANNOT** be called more than once.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +256,158 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+ * @user_script_message_reply: A #WebKitUserScriptMessageReply
+ * @reply_value: Reply value of the provided script message
+ *
+ * Provide the JS value reply to the script message sent in.
```
Reply to a script message with `@value`

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +256,158 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+ * @user_script_message_reply: A #WebKitUserScriptMessageReply
+ * @error_message: An error message to return as specified by the user's script message
+ *
+ * Provide the error message specified in the user provided script message.
```
Reply to a script message with and error message

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -320,6 +503,42 @@ void webkit_user_content_manager_unregister_script_message_handler(WebKitUserCon
+ *
+ * The registered handler can be unregistered by using
+ * #webkit_user_content_manager_unregister_script_message_handler and
+ * #webkit_user_content_manager_unregister_script_message_handler_in_world.
```
#webkit_user_content_manager_unregister_script_message_handler -> webkit_user_content_manager_unregister_script_message_handler()
#webkit_user_content_manager_unregister_script_message_handler_in_world -> webkit_user_content_manager_unregister_script_message_handler_in_world()

carlosgcampos reviewed (changes requested):
Looks much better, but there are still a few things to change. Thanks!

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -104,6 +109,34 @@ static void webkit_user_content_manager_class_init(WebKitUserContentManagerClass
+    /**
+     * WebKitUserContentManager::script-message-with-reply-received:
+     * @manager: the #WebKitUserContentManager
+     * @script_message_reply: the #WebKitScriptMessageReply holding the value received from the JavaScript world.
```
This needs to be updated, the value received is @js_result and @script_message_reply is used only to send the reply

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -104,6 +109,34 @@ static void webkit_user_content_manager_class_init(WebKitUserContentManagerClass
+            static_cast<GSignalFlags>(G_SIGNAL_RUN_LAST | G_SIGNAL_DETAILED),
+            0, g_signal_accumulator_true_handled, nullptr,
+            g_cclosure_marshal_generic,
+            G_TYPE_BOOLEAN, 1,
```
G_TYPE_BOOLEAN -> G_TYPE_NONE
1 -> 2

aperezdc replied:
We need `G_TYPE_BOOLEAN` to be able to use `g_signal_accumulator_true_handled`, which allows a signal callback to claim that it has handled the signal and that the rest of connected callbacks should not run.

Given that a script message request can only be replied or errored *only once* (and the API functions assert that on entry), we don't want multiple callbacks trying to send replies, therefore we need this in order to stop signal emission and avoid e.g. multiple calls to `webkit_user_script_message_reply_return_value()`.

aperezdc replied:
(Changing `1`→`2` is still needed, of course, now that the signal has one more argument.)

carlosgcampos replied:
true handled is useful when we have a default signal handler in the class, to prevent the default one from being invoked if already handled by the user. But user connecting twice to the same signal is just a programmer error, I would say.

mcatanzaro replied:
> true handled is useful when we have a default signal handler in the class, to prevent the default one from being invoked if already handled by the user. But user connecting twice to the same signal is just a programmer error, I would say.

I would use `g_signal_accumulator_true_handled` anyway. It's an extra level of safety and there's no downside.

aperezdc replied:
I don't see how connecting more than one handler to the same signal can be a programmer error. If something, it sounds like an arbitrary and unnecessary limitation&mdash;which GObject itself does not impose, BTW.

One may want to connect a callback that does logging and returns `FALSE` so other callback functions get a chance of doing actual work, or I may want to use `g_signal_connect_after()` with a callback that does `g_assert_not_reached()` to help debugging in some case where my code expects some other code to have handled a message, or I may want to have a few callbacks which handle different kinds of messages each, and have a different set of them connected depending on some configuration... and I could continue coming up with possible ways in which being able to `return TRUE` from a callback to stop signal propagation is useful.

Last but not least, we do have other APIs which follow the same pattern, so I think if only because of that it's worth keeping the accumulator and boolean return.

aperezdc replied:
FWIW, I tend to favor APIs which follow the “provide mechanisms, not policies” paradigm, meaning that it tends to be better to allow people to decide how things better fit what their programs need, than to impose a strict policy on how things work&mdash;of course, we need to specify _some_ policies (like “always use the public embedding API from the same thread inside a `GMainLoop`”) because there may not be a way around them... but I think in this case we don't need to force a single way of doing things.

mcatanzaro replied:
E.g. Epiphany does this for the `decide-policy` signal; it handles different policy decision types in different places.

carlosgcampos replied:
You are both right, and now I've realized that we need to make it true handled to allow to handle the reply asynchronously. We should add the return value to the documentation and include that to handle it async you have to take a ref to the reply and return TRUE.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +257,140 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+/**
+ * WebKitScriptMessageReply: (ref-func webkit_script_message_reply_ref) (unref-func webkit_script_message_reply_unref)
+ *
+ * A helper class that sends user provided script message to the signal handler.
```
A reply for a script message received.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +257,140 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+ * Since: 2.40
+ */
+struct _WebKitScriptMessageReply {
+    _WebKitScriptMessageReply(WebCore::SerializedScriptValue& serializedScriptValue, WTF::Function<void(API::SerializedScriptValue*, const String&)>&& completionHandler)
```
serializedScriptValue is unused now

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +257,140 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+    }
+}
+
+WebKitScriptMessageReply* webKitScriptMessageReplyCreate(WebCore::SerializedScriptValue& serializedScriptValue, WTF::Function<void(API::SerializedScriptValue*, const String&)>&& completionHandler)
```
serializedScriptValue is no longer needed

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -242,18 +403,23 @@ class ScriptMessageClientGtk final : public WebScriptMessageHandler::Client {
     {
+        WebKitScriptMessageReply* message = webKitScriptMessageReplyCreate(serializedScriptValue, WTFMove(completionHandler));
+        gboolean resultValue;
+        g_signal_emit(m_manager, signals[SCRIPT_MESSAGE_WITH_REPLY_RECEIVED], m_handlerName, message, &resultValue);
```
You should build the WebKitJavascriptResult and pass it to the signal

carlosgcampos replied:
Also the resultValue is unused, there's no reason for the signal to return a boolean

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContent.h.in:
```diff
@@ -23,6 +23,7 @@
 #define WebKitUserContent_h
 
 #include <glib-object.h>
+#include <jsc/jsc.h>
```
This shouldn't be needed

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManagerPrivate.h:
```diff
@@ -20,9 +20,13 @@
 #include "WebKitUserContentManager.h"
 #include "WebUserContentControllerProxy.h"
+#include <WebCore/UserScript.h>
+#include <WebCore/UserStyleSheet.h>
```
Why are these needed here?

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +261,140 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+
+    void sendValue(JSCValue* value)
+    {
+        JSCContext* jscContext = SharedJavascriptContext::singleton().getOrCreateContext();
```
After commit https://github.com/WebKit/WebKit/commit/4041ec7696c53b8f7e64d3046d4f7ae3d180cf35 you can now change this to be:

```suggestion
        JSCContext* jscContext = API::SerializedScriptValue::sharedJSCContext();
```

siwei-li replied:
Thank you for the suggestions! A dumb question: do I have to include the 4041ec7 commit in the forked repo for pull request in order to pass the build test?

mcatanzaro replied:
Yeah, you'll want to rebase.

siwei-li replied:
Encountering a problem here: after replacing `SharedJavascriptContext::singleton().getOrCreateContext()` to `API::SerializedScriptValue::sharedJSCContext()` in Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp, ld.lld complains about "undefined symbol", when "APISerializedScriptValue.h" is already included. 
May I ask what's going wrong here?

mcatanzaro replied:
Looks like it is not exported... try this (untested):

```
diff --git a/Source/WebKit/Shared/API/APISerializedScriptValue.h b/Source/WebKit/Shared/API/APISerializedScriptValue.h
index 8fb9ba26ecac..ddb67f89af1e 100644
--- a/Source/WebKit/Shared/API/APISerializedScriptValue.h
+++ b/Source/WebKit/Shared/API/APISerializedScriptValue.h
@@ -73,7 +73,7 @@ public:
 #endif
 
 #if USE(GLIB)
-    static JSCContext* sharedJSCContext();
+    WK_EXPORT static JSCContext* sharedJSCContext();
     static GRefPtr<JSCValue> deserialize(WebCore::SerializedScriptValue&);
     static RefPtr<SerializedScriptValue> createFromGVariant(GVariant*);
 #endif
 ```

siwei-li replied:
Thank you! Now it's working

siwei-li replied:
After this replacement of shared context, though, the script message handler is failing the unit test locally; I'm not sure why this is the case?

siwei-li replied:
Also I'm not quite certain about the reason why previously the passed tests in a local environment were unable to pass the EWS bot's APITests... Is there some sort of caching mechanism at work here that tricked me into believing the changed code was working? Any input would be appreciated 🙏

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -224,12 +261,140 @@ void webkit_user_content_manager_remove_all_scripts(WebKitUserContentManager* ma
+    ~_WebKitScriptMessageReply()
+    {
+        if (completionHandler) {
+            auto value = adoptGRef(jsc_value_new_undefined(SharedJavascriptContext::singleton().getOrCreateContext()));
```
Same here, after commit https://github.com/WebKit/WebKit/commit/4041ec7696c53b8f7e64d3046d4f7ae3d180cf35 this can be:

```suggestion
            auto value = adoptGRef(jsc_value_new_undefined(API::SerializedScriptValue::sharedJSCContext()));
```

aperezdc reviewed (commented):
Just a couple of small comments with a small simplification 😉

siwei-li committed:
[GTK][WPE] Support asynchronously returning values from user script messages
https://bugs.webkit.org/show_bug.cgi?id=243492

Reviewed by Carlos Garcia Campos.

Expose WebKit's functionality of responding asynchronously
to user script messages in the public API, by means of a
returned Promise object.

A new helper class "WebKitUserScriptMessgeReply" is added to
Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp;
it holds the returned JS value or the error message from executing
the user script message, which will be sent to the user by the
"didPostMessageWithAsyncReply()" method in the "ScriptMessageClientGtk"
class.

If asynchronous handling is wished for, the new
"webkit_user_content_manager_register_script_message_handler_with_reply()"
method should be used to register the message handler, its associated signal
being "WebKitUserContentManager::script-message-with-reply-received".
The user should ideally provide a callback reply in the script message,
otherwise an undefined value will be returned.

* Source/WebKit/Shared/API/APISerializedScriptValue.h:
* Source/WebKit/UIProcess/API/glib/APISerializedScriptValueGLib.cpp:
(API::SerializedScriptValue::createFromJSCValue):
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
(_WebKitScriptMessageReply::_WebKitScriptMessageReply):
(_WebKitScriptMessageReply::sendValue):
(_WebKitScriptMessageReply::sendErrorMessage):
(_WebKitScriptMessageReply::~_WebKitScriptMessageReply):
(webkit_script_message_reply_ref):
(webkit_script_message_reply_unref):
(webKitScriptMessageReplyCreate):
(webkit_script_message_reply_return_value):
(webkit_script_message_reply_return_error_message):
(webkit_user_content_manager_register_script_message_handler):
(webkit_user_content_manager_register_script_message_handler_with_reply):
(webkit_user_content_manager_register_script_message_handler_in_world):
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManagerPrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
(UserScriptMessageTest::UserScriptMessageTest):
(UserScriptMessageTest::~UserScriptMessageTest):
(UserScriptMessageTest::registerHandler):
(UserScriptMessageTest::scriptMessageReceived):
(UserScriptMessageTest::waitUntilMessageReceived):
(UserScriptMessageTest::asyncScriptMessageReceived):
(UserScriptMessageTest::waitUntilPromiseResolved):
(UserScriptMessageTest::runAsyncJavaScriptFinished):
(UserScriptMessageTest::postMessageAndWaitForPromiseResolved):
(testUserContentManagerScriptMessageWithReplyReceived):
(beforeAll):

Canonical link: https://commits.webkit.org/257016@main
